src/main/java/ch/njol/skript/lang/function/FunctionRegistry.java

```diff
@@ -0,0 +1,701 @@
+		if (!ns.identifiers.containsKey(provided.name)) {
+			Skript.debug("No signatures named '%s' exist in the '%s' namespace", provided.name, namespace.name);
+			return new Retrieval<>(RetrievalResult.NOT_REGISTERED, null, null);
+		}
+
+		Set<FunctionIdentifier> candidates = candidates(provided, ns.identifiers.get(provided.name));
+		if (candidates.isEmpty()) {
+			Skript.debug("Failed to find a signature for '%s'", provided.name);
+			return new Retrieval<>(RetrievalResult.NOT_REGISTERED, null, null);
+		} else if (candidates.size() == 1) {
+			if (Skript.debug()) {
+				Skript.debug("Matched signature for '%s': %s",
+					provided.name, ns.signatures.get(candidates.stream().findAny().orElse(null)));
+			}
+			return new Retrieval<>(RetrievalResult.EXACT,
+				ns.signatures.get(candidates.stream().findAny().orElse(null)),
+				null);
+		} else {
+			if (Skript.debug()) {
+				String options = candidates.stream().map(Record::toString).collect(Collectors.joining(", "));
+				Skript.debug("Failed to match an exact signature for '%s'", provided.name);
+				Skript.debug("Identifier: %s", provided);
+				Skript.debug("Options: %s", options);
+			}
+			return new Retrieval<>(RetrievalResult.AMBIGUOUS,
+				null,
+				candidates.stream()
+					.map(FunctionIdentifier::args)
+					.toArray(Class<?>[][]::new));
+		}
+	}
+
+	/**
+	 * Returns a list of candidates for the provided function identifier.
+	 *
+	 * @param provided The provided function.
+	 * @param existing The existing functions with the same name.
+	 * @return A list of candidates for the provided function.
+	 */
+	private static @NotNull Set<FunctionIdentifier> candidates(
+		@NotNull FunctionIdentifier provided,
+		Set<FunctionIdentifier> existing
+	) {
+		Set<FunctionIdentifier> candidates = new HashSet<>();
+
+		candidates:
+		for (FunctionIdentifier candidate : existing) {
+			// by this point, all candidates have matching names
+
+			// if we have no provided arguments, we can match any function
+			// we can skip the rest of the checks
+			if (provided.args == null || provided.args.length == 0) {
+				candidates.add(candidate);
+				continue;
+			}
+
+			// if argument counts are not possible, skip
+			if (provided.args.length > candidate.args.length
+				|| provided.args.length < candidate.minArgCount) {
+				continue;
+			}
+
+			// if the types of the provided arguments do not match the candidate arguments, skip
+			for (int i = 0; i < provided.args.length; i++) {
+				if (!Converters.converterExists(provided.args[i], candidate.args[i])) {
+					continue candidates;
+				}
+			}
+
+			candidates.add(candidate);
+		}
+
+		if (candidates.isEmpty()) {
+			return Set.of();
+		} else if (candidates.size() == 1 || provided.args == null || provided.args.length == 0) {
+			// if there is only one candidate,
+			// or we should match any function if provided.args == null || provided.args.length == 0,
+			// then return without trying to convert
+			return candidates;
+		}
```

Comment:
```suggestion
		if (candidates.size() <= 1 || provided.args == null || provided.args.length == 0) {
			// if there are no candidates, we return early
			// if there is only one candidate,
			// or we should match any function if provided.args == null || provided.args.length == 0,
			// then return without trying to convert
			return candidates;
		}
```
Simplification, and `Set.of()` is unmodifiable which leads to undocumented conditions on whether the returned set is modifiable.